File: generic-utilities/src/Validator.ts
/**
 * Validation result type - success with data or failure with errors
 */
export type ValidationResult<T> = {
    success: true;
    data: T;
} | {
    success: false;
    errors: string[];
};

/**
 * Validator function type
 */
export type Validator<T> = (value: unknown) => ValidationResult<T>;

/**
 * Create a validator for string values
 * @returns Validator that checks if value is a string
 */
export function string(): Validator<string> {
    return (value: unknown): ValidationResult<string> => {
        if (typeof value === 'string') {
            return { success: true, data: value };
        }
        return { success: false, errors: ['Expected string'] };
    };
}

/**
 * Create a validator for number values
 * @returns Validator that checks if value is a number
 */
export function number(): Validator<number> {
    return (value: unknown): ValidationResult<number> => {
        if (typeof value === 'number' && !isNaN(value)) {
            return { success: true, data: value };
        }
        return { success: false, errors: ['Expected number'] };
    };
}

/**
 * Create a validator for boolean values
 * @returns Validator that checks if value is a boolean
 */
export function boolean(): Validator<boolean> {
    return (value: unknown): ValidationResult<boolean> => {
        if (typeof value === 'boolean') {
            return { success: true, data: value };
        }
        return { success: false, errors: ['Expected boolean'] };
    };
}

/**
 * Create a validator for array values with item validation
 * @template T - Type of array items
 * @param itemValidator - Validator for each array item
 * @returns Validator that checks if value is an array with valid items
 */
export function array<T>(itemValidator: Validator<T>): Validator<T[]> {
    return (value: unknown): ValidationResult<T[]> => {
        if (!Array.isArray(value)) {
            return { success: false, errors: ['Expected array'] };
        }

        const results: T[] = [];
        const errors: string[] = [];

        for (let i = 0; i < value.length; i++) {
            const result = itemValidator(value[i]);
            if (result.success) {
                results.push(result.data);
            } else {
                errors.push(`Item at index ${i}: ${result.errors.join(', ')}`);
            }
        }

        if (errors.length > 0) {
            return { success: false, errors };
        }

        return { success: true, data: results };
    };
}

/**
 * Create a validator for object values with schema validation
 * @template T - Object type
 * @param schema - Schema defining validators for each property
 * @returns Validator that checks if value matches the schema
 */
export function object<T extends Record<string, unknown>>(
    schema: { [K in keyof T]: Validator<T[K]> }
): Validator<T> {
    return (value: unknown): ValidationResult<T> => {
        if (typeof value !== 'object' || value === null || Array.isArray(value)) {
            return { success: false, errors: ['Expected object'] };
        }

        const obj = value as Record<string, unknown>;
        const result: Partial<T> = {};
        const errors: string[] = [];

        for (const key in schema) {
            const validator = schema[key];
            const fieldValue = obj[key];
            const fieldResult = validator(fieldValue);

            if (fieldResult.success) {
                result[key] = fieldResult.data;
            } else {
                errors.push(`Field '${key}': ${fieldResult.errors.join(', ')}`);
            }
        }

        if (errors.length > 0) {
            return { success: false, errors };
        }

        return { success: true, data: result as T };
    };
}

/**
 * Create a validator for optional values
 * @template T - Value type
 * @param validator - Validator for the value when present
 * @returns Validator that allows undefined or validates with the given validator
 */
export function optional<T>(validator: Validator<T>): Validator<T | undefined> {
    return (value: unknown): ValidationResult<T | undefined> => {
        if (value === undefined) {
            return { success: true, data: undefined };
        }
        return validator(value);
    };
}

/**
 * Create a validator for nullable values
 * @template T - Value type
 * @param validator - Validator for the value when not null
 * @returns Validator that allows null or validates with the given validator
 */
export function nullable<T>(validator: Validator<T>): Validator<T | null> {
    return (value: unknown): ValidationResult<T | null> => {
        if (value === null) {
            return { success: true, data: null };
        }
        return validator(value);
    };
}

/**
 * Create a validator for literal values
 * @template T - Literal type
 * @param literal - The expected literal value
 * @returns Validator that checks if value equals the literal
 */
export function literal<T extends string | number | boolean>(literal: T): Validator<T> {
    return (value: unknown): ValidationResult<T> => {
        if (value === literal) {
            return { success: true, data: value as T };
        }
        return { success: false, errors: [`Expected literal value: ${literal}`] };
    };
}

/**
 * Create a validator for union types
 * @template T - Union type
 * @param validators - Array of validators to try
 * @returns Validator that succeeds if any validator succeeds
 */
export function union<T>(...validators: Validator<T>[]): Validator<T> {
    return (value: unknown): ValidationResult<T> => {
        const errors: string[] = [];

        for (const validator of validators) {
            const result = validator(value);
            if (result.success) {
                return result;
            }
            errors.push(...result.errors);
        }

        return { success: false, errors: ['No union variant matched', ...errors] };
    };
}

/**
 * Create a custom validator with a predicate function
 * @template T - Value type
 * @param validator - Base validator
 * @param predicate - Additional validation predicate
 * @param errorMessage - Error message if predicate fails
 * @returns Validator with custom validation
 */
export function refine<T>(
    validator: Validator<T>,
    predicate: (value: T) => boolean,
    errorMessage: string
): Validator<T> {
    return (value: unknown): ValidationResult<T> => {
        const result = validator(value);
        if (!result.success) {
            return result;
        }

        if (!predicate(result.data)) {
            return { success: false, errors: [errorMessage] };
        }

        return result;
    };
}

File: golang-utilities/generic-utils/validator.go
package genericutils

import (
	"fmt"
	"reflect"
)

// ValidationResult represents the result of a validation
type ValidationResult[T any] struct {
	Success bool
	Data    T
	Errors  []string
}

// Validator is a function that validates a value
type Validator[T any] func(value interface{}) ValidationResult[T]

// String creates a validator for string values
func String() Validator[string] {
	return func(value interface{}) ValidationResult[string] {
		if str, ok := value.(string); ok {
			return ValidationResult[string]{
				Success: true,
				Data:    str,
				Errors:  nil,
			}
		}
		return ValidationResult[string]{
			Success: false,
			Data:    "",
			Errors:  []string{"Expected string"},
		}
	}
}

// Number creates a validator for numeric values (int, int64, float64)
func Number() Validator[float64] {
	return func(value interface{}) ValidationResult[float64] {
		switch v := value.(type) {
		case int:
			return ValidationResult[float64]{Success: true, Data: float64(v)}
		case int32:
			return ValidationResult[float64]{Success: true, Data: float64(v)}
		case int64:
			return ValidationResult[float64]{Success: true, Data: float64(v)}
		case float32:
			return ValidationResult[float64]{Success: true, Data: float64(v)}
		case float64:
			return ValidationResult[float64]{Success: true, Data: v}
		default:
			return ValidationResult[float64]{
				Success: false,
				Data:    0,
				Errors:  []string{"Expected number"},
			}
		}
	}
}

// Integer creates a validator for integer values
func Integer() Validator[int] {
	return func(value interface{}) ValidationResult[int] {
		switch v := value.(type) {
		case int:
			return ValidationResult[int]{Success: true, Data: v}
		case int32:
			return ValidationResult[int]{Success: true, Data: int(v)}
		case int64:
			return ValidationResult[int]{Success: true, Data: int(v)}
		case float64:
			if v == float64(int(v)) {
				return ValidationResult[int]{Success: true, Data: int(v)}
			}
		}
		return ValidationResult[int]{
			Success: false,
			Data:    0,
			Errors:  []string{"Expected integer"},
		}
	}
}

// Boolean creates a validator for boolean values
func Boolean() Validator[bool] {
	return func(value interface{}) ValidationResult[bool] {
		if b, ok := value.(bool); ok {
			return ValidationResult[bool]{
				Success: true,
				Data:    b,
				Errors:  nil,
			}
		}
		return ValidationResult[bool]{
			Success: false,
			Data:    false,
			Errors:  []string{"Expected boolean"},
		}
	}
}

// Array creates a validator for array/slice values with item validation
func Array[T any](itemValidator Validator[T]) Validator[[]T] {
	return func(value interface{}) ValidationResult[[]T] {
		rv := reflect.ValueOf(value)
		if rv.Kind() != reflect.Slice && rv.Kind() != reflect.Array {
			return ValidationResult[[]T]{
				Success: false,
				Data:    nil,
				Errors:  []string{"Expected array"},
			}
		}

		results := make([]T, 0, rv.Len())
		errors := make([]string, 0)

		for i := 0; i < rv.Len(); i++ {
			item := rv.Index(i).Interface()
			result := itemValidator(item)
			if result.Success {
				results = append(results, result.Data)
			} else {
				errors = append(errors, fmt.Sprintf("Item at index %d: %v", i, result.Errors))
			}
		}

		if len(errors) > 0 {
			return ValidationResult[[]T]{
				Success: false,
				Data:    nil,
				Errors:  errors,
			}
		}

		return ValidationResult[[]T]{
			Success: true,
			Data:    results,
			Errors:  nil,
		}
	}
}

// Object creates a validator for struct/map objects
func Object[T any](validators map[string]Validator[interface{}]) Validator[T] {
	return func(value interface{}) ValidationResult[T] {
		rv := reflect.ValueOf(value)
		var zero T

		// Handle maps
		if rv.Kind() == reflect.Map {
			errors := make([]string, 0)

			for field, validator := range validators {
				mapValue := rv.MapIndex(reflect.ValueOf(field))
				if !mapValue.IsValid() {
					errors = append(errors, fmt.Sprintf("Field '%s': missing", field))
					continue
				}

				result := validator(mapValue.Interface())
				if !result.Success {
					errors = append(errors, fmt.Sprintf("Field '%s': %v", field, result.Errors))
				}
			}

			if len(errors) > 0 {
				return ValidationResult[T]{
					Success: false,
					Data:    zero,
					Errors:  errors,
				}
			}

			// Try to convert to T
			if typedValue, ok := value.(T); ok {
				return ValidationResult[T]{
					Success: true,
					Data:    typedValue,
					Errors:  nil,
				}
			}
		}

		// Handle structs
		if rv.Kind() == reflect.Struct || (rv.Kind() == reflect.Ptr && rv.Elem().Kind() == reflect.Struct) {
			if rv.Kind() == reflect.Ptr {
				rv = rv.Elem()
			}

			errors := make([]string, 0)

			for field := range validators {
				fieldValue := rv.FieldByName(field)
				if !fieldValue.IsValid() {
					errors = append(errors, fmt.Sprintf("Field '%s': not found", field))
				}
			}

			if len(errors) > 0 {
				return ValidationResult[T]{
					Success: false,
					Data:    zero,
					Errors:  errors,
				}
			}

			// Try to convert to T
			if typedValue, ok := value.(T); ok {
				return ValidationResult[T]{
					Success: true,
					Data:    typedValue,
					Errors:  nil,
				}
			}
		}

		return ValidationResult[T]{
			Success: false,
			Data:    zero,
			Errors:  []string{"Expected object"},
		}
	}
}

// Optional creates a validator for optional values
func Optional[T any](validator Validator[T]) Validator[*T] {
	return func(value interface{}) ValidationResult[*T] {
		if value == nil {
			return ValidationResult[*T]{
				Success: true,
				Data:    nil,
				Errors:  nil,
			}
		}

		result := validator(value)
		if result.Success {
			return ValidationResult[*T]{
				Success: true,
				Data:    &result.Data,
				Errors:  nil,
			}
		}

		return ValidationResult[*T]{
			Success: false,
			Data:    nil,
			Errors:  result.Errors,
		}
	}
}

// Literal creates a validator for specific literal values
func Literal[T comparable](literal T) Validator[T] {
	return func(value interface{}) ValidationResult[T] {
		if v, ok := value.(T); ok && v == literal {
			return ValidationResult[T]{
				Success: true,
				Data:    v,
				Errors:  nil,
			}
		}
		return ValidationResult[T]{
			Success: false,
			Data:    literal,
			Errors:  []string{fmt.Sprintf("Expected literal value: %v", literal)},
		}
	}
}

// Union creates a validator that tries multiple validators
func Union[T any](validators ...Validator[T]) Validator[T] {
	return func(value interface{}) ValidationResult[T] {
		errors := make([]string, 0)

		for _, validator := range validators {
			result := validator(value)
			if result.Success {
				return result
			}
			errors = append(errors, result.Errors...)
		}

		var zero T
		return ValidationResult[T]{
			Success: false,
			Data:    zero,
			Errors:  append([]string{"No union variant matched"}, errors...),
		}
	}
}

// Refine creates a validator with an additional predicate
func Refine[T any](validator Validator[T], predicate func(T) bool, errorMessage string) Validator[T] {
	return func(value interface{}) ValidationResult[T] {
		result := validator(value)
		if !result.Success {
			return result
		}

		if !predicate(result.Data) {
			return ValidationResult[T]{
				Success: false,
				Data:    result.Data,
				Errors:  []string{errorMessage},
			}
		}

		return result
	}
}

// MinLength creates a string validator with minimum length
func MinLength(min int) Validator[string] {
	return Refine(String(), func(s string) bool {
		return len(s) >= min
	}, fmt.Sprintf("String must be at least %d characters", min))
}

// MaxLength creates a string validator with maximum length
func MaxLength(max int) Validator[string] {
	return Refine(String(), func(s string) bool {
		return len(s) <= max
	}, fmt.Sprintf("String must be at most %d characters", max))
}

// Min creates a number validator with minimum value
func Min(min float64) Validator[float64] {
	return Refine(Number(), func(n float64) bool {
		return n >= min
	}, fmt.Sprintf("Number must be at least %v", min))
}

// Max creates a number validator with maximum value
func Max(max float64) Validator[float64] {
	return Refine(Number(), func(n float64) bool {
		return n <= max
	}, fmt.Sprintf("Number must be at most %v", max))
}

// Email creates a basic email validator
func Email() Validator[string] {
	return Refine(String(), func(s string) bool {
		// Simple email validation
		return len(s) > 0 && containsChar(s, '@') && containsChar(s, '.')
	}, "Invalid email format")
}

func containsChar(s string, c rune) bool {
	for _, ch := range s {
		if ch == c {
			return true
		}
	}
	return false
}
